### DISCUSSION TITLE
require is not defined in ES module scope after Build

### DISCUSSION THREAD

**Discussion by olegbelov21** [Lillifee/raspiCam#35] [Q&A]
Hello, everyone,

I wanted to try to compile the project myself on the Raspberry pi (because I wanted to change the port from 8000 to 80 in the file raspiCam/src/server/main.ts).

The project compiles:

> ```
> git clone https://github.com/Lillifee/raspiCam.git
> 
> cd raspiCam/
> 
> npm install
> npm run build
> ```
> The bundled package should appear in the **build** folder.

but I get the following error when running **node server.js**

ReferenceError: require is not defined in ES module scope, you can use import instead
This file is being treated as an ES module because it has a '.js' file extension and '/home/pi/projects/raspiCam/package.json' contains "type": "module". To treat it as a CommonJS script, rename it to use the '.cjs' file extension.
    at 3685 (file:///home/pi/projects/raspiCam/build/server.js:2:195430)
    at i (file:///home/pi/projects/raspiCam/build/server.js:2:464533)
    at file:///home/pi/projects/raspiCam/build/server.js:2:464895
    at file:///home/pi/projects/raspiCam/build/server.js:2:484115
    at file:///home/pi/projects/raspiCam/build/server.js:2:484119
    at ModuleJob.run (node:internal/modules/esm/module_job:194:25)

Node.js v18.13.0

I have already tried deleting or changing "type": "module" in the package.json.
does anyone have an idea? 

Best regards

**Comment by Lillifee** [comment]
Hey @olegbelov21 

Thanks for testing it on the Raspberry pi.
I recently switched from commonJs to the ES module and have some problems with it. (eg. I have to use a beta of styled components etc.) 

I usually build it on my pc and copy the build results to the raspberry pi, and it seems there is a problem. 
If you run the server.js from the ./build folder, it somehow uses the package.json from the parent folder to identify the module type.


As a quick fix, you could copy/move the build folder to another location and run the server from there.
```
git clone https://github.com/Lillifee/raspiCam.git

cd raspiCam/

npm install --force
npm run build 

cp -r ./build ../raspiCamBuild
```

Or you could remove the package.json after the build.
```
git clone https://github.com/Lillifee/raspiCam.git

cd raspiCam/

npm install --force
npm run build 

rm package.json
```

I will also add the port as a startup argument with the next release.
```
node server.js -p 80
```

I'm not familiar with the ES module and know the pitfalls, especially with webpack. I'll try to find a solution. 
Sorry for that.

**Reply by Lillifee** [reply]
Hey @olegbelov21,

I created a new release with the command line argument -p or --port and the new pi camera 3 focus settings.

https://github.com/Lillifee/raspiCam/releases/tag/v2.0.1

**Reply by olegbelov21** [reply]
> Hey @olegbelov21,
> 
> I created a new release with the command line argument -p or --port and the new pi camera 3 focus settings.
> 
> https://github.com/Lillifee/raspiCam/releases/tag/v2.0.1

Hey @Lillifee ,

Thank you for the fast answer.
Using the port as an agrument is a good idea.

I wanted to try to compile the project anyway. 
I tried deleting package.json or moving the build folder. Then I have this error:

Error: node loader:
Error: /home/pi/projects/build_raspi/build//epoll.node: wrong ELF class: ELFCLASS32
     at 8698 (/home/pi/projects/build_raspi/build/server.js:2:195121)
     at i (/home/pi/projects/build_raspi/build/server.js:2:464533)
     at /home/pi/projects/build_raspi/build/server.js:2:31056
     at 9728 (/home/pi/projects/build_raspi/build/server.js:2:31164)
     at i (/home/pi/projects/build_raspi/build/server.js:2:464533)
     at 8640 (/home/pi/projects/build_raspi/build/server.js:2:153486)
     at i (/home/pi/projects/build_raspi/build/server.js:2:464533)
     at /home/pi/projects/build_raspi/build/server.js:2:464914
     at /home/pi/projects/build_raspi/build/server.js:2:484115
     at Object.<anonymous> (/home/pi/projects/build_raspi/build/server.js:2:484119)

Node.js v18.13.0


Thank you for the adaptation in version 2.0.1 I just tried the compiled version and I also get the error as above.

wget https://github.com/Lillifee/raspiCam/releases/download/v2.0.1/raspiCam.tar.gz && tar -xvzf raspiCam.tar.gz
node server.js _**-> same error as above**_

but the version 2.0.0 works on the same system
wget https://github.com/Lillifee/raspiCam/releases/download/v2.0.0/raspiCam.tar.gz && tar -xvzf raspiCam.tar.gz
node server.js


Best regards

**Reply by Lillifee** [reply]
Hey @olegbelov21 

thats really strange, i just tried it on my raspberry 4, and the v2.0.1 worked fine for me. 
We had a similar issue with epoll with a different node version. https://github.com/Lillifee/raspiCam/issues/33

wrong ELF class: ELFCLASS32 sound like you are running it on a 64bit version but i dont understand how the v2.0.0 could work.
There was no change to the epoll.node binary in this realease.

Can you try to copy the epoll.node from the working 2.0.0 to the 2.0.1?  
You could also try to compile epoll node on your raspberry pi using: 

```npm install epoll```

and copy the epoll node from ./node_modules/epoll/build/Release/epoll.node to the raspiCam folder.
I'm still searching for a solution to access the gpio's without a precompiled binary.

**Reply by olegbelov21** [reply]
> Hey @olegbelov21
> 
> thats really strange, i just tried it on my raspberry 4, and the v2.0.1 worked fine for me. We had a similar issue with epoll with a different node version. #33
> 
> wrong ELF class: ELFCLASS32 sound like you are running it on a 64bit version but i dont understand how the v2.0.0 could work. There was no change to the epoll.node binary in this realease.
> 
> Can you try to copy the epoll.node from the working 2.0.0 to the 2.0.1? You could also try to compile epoll node on your raspberry pi using:
> 
> `npm install epoll`
> 
> and copy the epoll node from ./node_modules/epoll/build/Release/epoll.node to the raspiCam folder. I'm still searching for a solution to access the gpio's without a precompiled binary.

hey @Lillifee ,

Thanks! yes i have 64 bit version.
recompiling and copying epoll worked. Recompiling the project and running it from the build folder works now too.

maybe as an addition to the README also for RPi 4 
```
npm install epoll && cp node_modules/epoll/build/Release/epoll.node .
```


i think without installing npm package it will be difficult to monitor a GPIO. Unless you implement a solution calling bash script files with GPIO access from your project.
